src/Tools/dotnet-dump/Analyzer.cs

```diff
@@ -25,45 +26,60 @@ public async Task<int> Analyze(FileInfo dump_path, string[] command)
-                if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) {
-                    await _commandProcessor.Parse("setsymbolserver -ms", _consoleProvider);
+            try
+            { 
+                DataTarget target = null;
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) {
+                    target = DataTarget.LoadCoreDump(dump_path.FullName);
+                }
+                else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) {
+                    target = DataTarget.LoadCrashDump(dump_path.FullName, CrashDumpReader.ClrMD);
+                }
+                else {
+                    throw new PlatformNotSupportedException($"Unsupported operating system: {RuntimeInformation.OSDescription}");
                 }
 
-                // Run the commands from the dotnet-dump command line
-                if (command != null)
+                using (target)
                 {
-                    foreach (string cmd in command)
+                    // Create common analyze context for commands
+                    var analyzeContext = new AnalyzeContext(_consoleProvider, target, _consoleProvider.Stop) {
+                        CurrentThreadId = unchecked((int)target.DataReader.EnumerateAllThreads().FirstOrDefault())
+                    };
+                    _commandProcessor.CommandContext = analyzeContext;
+
+                    // Automatically enable symbol server support on Linux and MacOS
+                    if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) {
+                        await _commandProcessor.Parse("setsymbolserver -ms", _consoleProvider);
+                    }
+
+                    // Run the commands from the dotnet-dump command line
+                    if (command != null)
                     {
-                        await _commandProcessor.Parse(cmd, _consoleProvider);
+                        foreach (string cmd in command)
+                        {
+                            await _commandProcessor.Parse(cmd, _consoleProvider);
+                        }
                     }
-                }
 
-                // Start interactive command line processing
-                await _consoleProvider.Start(async (string commandLine, CancellationToken cancellation) => {
-                    analyzeContext.CancellationToken = cancellation;
-                    await _commandProcessor.Parse(commandLine, _consoleProvider);
-                });
+                    // Start interactive command line processing
+                    await _consoleProvider.Start(async (string commandLine, CancellationToken cancellation) => {
+                        analyzeContext.CancellationToken = cancellation;
+                        await _commandProcessor.Parse(commandLine, _consoleProvider);
+                    });
+                }
+            }
+            catch (Exception ex) when 
+                (ex is ClrDiagnosticsException ||
```

Comment:
This form of error handling seems likely to catch unintended exceptions and when it does it will display unuseful error strings that lack context and are harder to debug. I still recommend using a specific exception type that is under the app's control to ensure you are only catching exactly the exceptions you intend to catch. You can always insert into particular places in code:

    try
    {
        // specific operation that might fail with IOException
    }
    catch(IOException e)
    {
        throw new AnalyzerException("Whatever user-visible error message is appropriate", e);
    }

Reply:
Unless I'm misunderstanding what you are saying or I don't understand the "when" option on the catch, this should only catch the exceptions listed.  This was mostly to get around that System.CommandLine exception catch that catches everything and prints the stack frame. It is really ugly for most exceptions.

Reply:
The scenario I am imagining is that somewhere your code calls into CLRMD, which calls into the framework, which calls deep into the bowels of some helper routine, and way down there it throws 'ArgumentException' (or any other exception in your list). Presumably that exception doesn't represent a user error, it represents a bug that we didn't anticipate. Instead of getting a crash/stacktrace we could debug at Microsoft, that exception is going to get caught here and then the user is going to see:

    $ dotnet dump analyze core123213
       Analyzing dump core123213:
       Value
    $

I'm suggesting that if the exceptions represent bugs, we probably do want an ugly stack trace because at least it is more easily debuggable. I don't imagine a user would be any happier if they saw the output above, and if they posted it on github as a bug report now the burden lands on you to figure out where it is failing with potentially no repro and no stack trace.

Reply:
Also fine if you want to postpone fixing this, but I think its going to be worth your while not to get those awkward bug reports that don't have stacks.

Reply:
You can customize the System.CommandLine exception behavior: 

https://github.com/dotnet/command-line-api/blob/master/src/System.CommandLine.Tests/UseExceptionHandlerTests.cs#L111

Reply:
What about removing IOException (replacing it with FileNotFound and DirectoryNotFound) and ArgumentException? 
    
    catch (Exception ex) when 
                (ex is ClrDiagnosticsException ||
                 ex is FileNotFoundException || 
                 ex is DirectoryNotFoundException || 
                 ex is UnauthorizedAccessException || 
                 ex is PlatformNotSupportedException || 
                 ex is InvalidDataException ||
                 ex is InvalidOperationException ||
                 ex is NotSupportedException)

These are the more specific exceptions that are thrown by either my code or CLRMD. They still may not be specific enough and end up with similar messages.

For all the exceptions that my code throws (PlatformNotSupportedException, InvalidOperationException, NotSupportedException), I could replace it with a custom one, but that doesn't help with the exceptions throw by CLRMD, etc. 

Another approach is not too catch any exceptions (in both Dumper.cs and Analyzer.cs) and live with the stack dump in all cases, I think this is really ugly for "common" exceptions like can't find the core dump file or operating system not supported.

Reply:
> I think this is really ugly for "common" exceptions like can't find the core dump file or operating system not supported.

I would suggest that divide exceptions into two groups:
1) Errors that we predict users will see - We spend some effort on that group to ensure the error message is reasonable. This doesn't necessarily require that we provide a custom error message for a huge number of different exceptions. We might just add some generic context and then pass through whatever the original error was:
"Error opening dump file {path}. Error {ex.Message} "
We can tailor the amount of effort we spend on error message clarity relative to how likely it is users will actually see those errors. If we guess wrong we can always refine with future incremental investments.

2) Errors that we did not anticipate at all - these are bugs and if it turns out users hit them then we probably want a stack trace to show up somewhere. It doesn't have to be on the console. For example maybe the experience is:

    $ dotnet dump analyze core123213
       Analyzing dump core123213:
       An unexpected error occurred with error message "Value"
       Diagnostic details for this error have been saved to error.txt.
   $



> I could replace it with a custom one, but that doesn't help with the exceptions throw by CLRMD, etc.

If you have a specific operation you are doing with CLRMD, and specific exceptions you expect that to throw you can always catch it and throw a new exception which provides the original exception as the inner exception. Its true that you may not be able to filter perfectly in this case, but wrapping also gives you the oportunity to create an error message that adds context. For example instead of FileNotFound exception message of "File not found", you could say "Unable to open dump file {path}. Error: {ex.Message}"